ui/src/main/java/com/xmartlabs/bigbang/ui/common/recyclerview/BaseRecyclerViewAdapter.java

```diff
@@ -0,0 +1,319 @@
+        .indexed()
+        .forEach(itemWithPosition ->
+            addItemWithoutNotifying(index + itemWithPosition.getFirst(), type, itemWithPosition.getSecond(), false));
+    addItemTypeIfNeeded(type);
+    notifyItemRangeInserted(index, getItemCount() - lastItemCount);
+    return true;
+  }
+
+  /**
+   * Adds items to the data for the recycler view and notifies any registered observers that the items has been added.
+   *
+   * @param type  The type of the items.
+   * @param items the items that will be the data for the recycler view.
+   * @return if item was successfully added.
+   */
+  @MainThread
+  protected <T extends RecycleItemType> boolean addItems(@NonNull T type, @Nullable List<?> items) {
+    if (CollectionHelper.isNullOrEmpty(items)) {
+      return false;
+    }
+    int lastItemCount = getItemCount();
+    Stream.ofNullable(items)
+        .forEach(item -> addItemWithoutNotifying(type, item, false));
+    addItemTypeIfNeeded(type);
+    if (lastItemCount == 0) {
+      notifyDataSetChanged();
+    } else {
+      notifyItemRangeInserted(lastItemCount, getItemCount() - lastItemCount);
+    }
+    return true;
+  }
+
+  /**
+   * Sets the items data for the recycler view and notifying any registered observers that the data set has
+   * changed. It uses a function that calculates the difference between the old and the new items
+   * in order to improve the update process.
+   *
+   * @param type                      Type of items.
+   * @param newItems                  The items tobe added.
+   * @param areItemsTheSameFunction   A function which checks that two items are the same.
+   * @param areContentTheSameFunction A function which checks that the content of two items are the same.
+   */
+  @SuppressWarnings("WeakerAccess")
+  protected <T extends RecycleItemType> void setItems(@NonNull T type, final @Nullable List<?> newItems,
+                                                      @NonNull BiFunction<Object, Object, Boolean> areItemsTheSameFunction,
+                                                      @NonNull BiFunction<Object, Object, Boolean> areContentTheSameFunction) {
+    if (CollectionHelper.isNullOrEmpty(newItems)) {
+      return;
+    }
+
+    if (updateElementsDisposable != null && !updateElementsDisposable.isDisposed()) {
+      updateElementsDisposable.dispose();
+    }
+    updateElementsDisposable = Single.fromCallable(() -> DiffUtil
+        .calculateDiff(getUpdateDiffCallback(newItems, areItemsTheSameFunction, areContentTheSameFunction)))
+        .subscribeOn(Schedulers.io())
+        .observeOn(AndroidSchedulers.mainThread())
+        .subscribe(diffResult -> {
+          items.clear();
+          Stream.of(newItems)
+              .forEach(item -> addItemWithoutNotifying(type, item, false));
+          addItemTypeIfNeeded(type);
+          diffResult.dispatchUpdatesTo(this);
+        });
+  }
+
+  @NonNull
+  private DiffUtil.Callback getUpdateDiffCallback(@NonNull final List<?> newItems,
```

Comment:
Yep I know, sometimes there isn't a nicer way. But it was just a thought, do what you think is better.